crypto/src/main/kotlin/web5/sdk/crypto/Ed25519.kt

```diff
@@ -0,0 +1,135 @@
+ * @property pubMulticodec A byte array representing the multicodec prefix for an Ed25519 public key.
+ * @property privMultiCodec A byte array representing the multicodec prefix for an Ed25519 private key.
+ */
+
+public object Ed25519 : KeyGenerator, Signer {
+  override val algorithm: Algorithm = JWSAlgorithm.EdDSA
+  override val keyType: KeyType = KeyType.OKP
+
+  public val pubMulticodec: ByteArray = Varint.encode(0xed)
+  public val privMultiCodec: ByteArray = Varint.encode(0x1300)
+
+  /**
+   * Generates a private key utilizing the Ed25519 algorithm.
+   *
+   * @param options (Optional) Additional options to control the key generation process.
+   * @return The generated private key in JWK format.
+   */
+  override fun generatePrivateKey(options: KeyGenOptions?): JWK {
+    return OctetKeyPairGenerator(Curve.Ed25519)
+      .keyIDFromThumbprint(true)
+      .keyUse(KeyUse.SIGNATURE)
+      .generate()
+      .toOctetKeyPair()
+  }
+
+  /**
+   * Derives the public key corresponding to a given private key.
+   *
+   * @param privateKey The private key in JWK format.
+   * @return The corresponding public key in JWK format.
+   */
+  override fun getPublicKey(privateKey: JWK): JWK {
+    require(privateKey is OctetKeyPair) { "private key must be an Octet Key Pair (kty: OKP)" }
+
+    return privateKey.toOctetKeyPair().toPublicJWK()
+  }
+
+  override fun privateKeyToBytes(privateKey: JWK): ByteArray {
+    validateKey(privateKey)
+
+    return privateKey.toOctetKeyPair().decodedD
+  }
+
+  override fun publicKeyToBytes(publicKey: JWK): ByteArray {
+    validateKey(publicKey)
+
+    return publicKey.toOctetKeyPair().decodedX
+  }
+
+  override fun bytesToPrivateKey(privateKeyBytes: ByteArray): JWK {
+    val privateKeyParameters = Ed25519PrivateKeyParameters(privateKeyBytes, 0)
+    val publicKeyBytes = privateKeyParameters.generatePublicKey().encoded
+
+    val base64UrlEncodedPrivateKey = Convert(privateKeyBytes).toBase64Url(padding = false)
+    val base64UrlEncodedPublicKey = Convert(publicKeyBytes).toBase64Url(padding = false)
+
+    return OctetKeyPair.Builder(Curve.Ed25519, Base64URL(base64UrlEncodedPublicKey))
+      .algorithm(algorithm)
+      .keyIDFromThumbprint()
+      .d(Base64URL(base64UrlEncodedPrivateKey))
+      .keyUse(KeyUse.SIGNATURE)
+      .build()
+  }
+
+  override fun bytesToPublicKey(publicKeyBytes: ByteArray): JWK {
+    val base64UrlEncodedPublicKey = Convert(publicKeyBytes).toBase64Url(padding = false)
+
+    return OctetKeyPair.Builder(Curve.Ed25519, Base64URL(base64UrlEncodedPublicKey))
+      .algorithm(algorithm)
+      .keyIDFromThumbprint()
+      .keyUse(KeyUse.SIGNATURE)
+      .build()
+  }
+
+  override fun sign(privateKey: JWK, payload: Payload, options: SignOptions?): String {
+    val jwsHeader = JWSHeader.Builder(JWSAlgorithm.EdDSA)
+      .keyID(privateKey.keyID)
+      .build()
+
+    val jws = JWSObject(jwsHeader, payload)
+    val signer = Ed25519Signer(privateKey as OctetKeyPair)
```

Comment:
Is this meant to sign only with Ed25519?

Reply:
yep it's in a file named `Ed25519`

Reply:
🤦